python/ray/experimental/iter.py

```diff
@@ -0,0 +1,556 @@
+                 timeout: int = None,
+                 name=None):
+        self.base_iterator = base_iterator
+        self.built_iterator = None
+        self.local_transforms = local_transforms or []
+        self.timeout = timeout
+        if name is None:
+            name = "LocalIterator[?]"
+        self.name = name
+
+    def _build_once(self):
+        if self.built_iterator is None:
+            it = iter(self.base_iterator(self.timeout))
+            for fn in self.local_transforms:
+                it = fn(it)
+            self.built_iterator = it
+
+    def __iter__(self):
+        self._build_once()
+        return self.built_iterator
+
+    def __next__(self):
+        self._build_once()
+        return next(self.built_iterator)
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return self.name
+
+    def for_each(self, fn: Callable[[T], T]) -> "LocalIterator[T]":
+        def apply_foreach(it):
+            for item in it:
+                if isinstance(item, YieldIterator):
+                    yield item
+                else:
+                    yield fn(item)
+
+        return LocalIterator(
+            self.base_iterator,
+            self.local_transforms + [apply_foreach],
+            name=self.name + ".for_each()")
+
+    def filter(self, fn: Callable[[T], bool]) -> "LocalIterator[T]":
+        def apply_filter(it):
+            for item in it:
+                if isinstance(item, YieldIterator):
+                    yield item
+                elif fn(item):
+                    yield item
+
+        return LocalIterator(
+            self.base_iterator,
+            self.local_transforms + [apply_filter],
+            name=self.name + ".filter()")
+
+    def batch(self, n: int) -> "LocalIterator[List[T]]":
+        def apply_batch(it):
+            batch = []
+            for item in it:
+                if isinstance(item, YieldIterator):
+                    yield item
+                else:
+                    batch.append(item)
+                    if len(batch) >= n:
+                        yield batch
+                        batch = []
+            if batch:
+                yield batch
+
+        return LocalIterator(
+            self.base_iterator,
+            self.local_transforms + [apply_batch],
+            name=self.name + ".batch({})".format(n))
+
+    def flatten(self) -> "LocalIterator[T[0]]":
+        def apply_flatten(it):
+            for item in it:
+                if isinstance(item, YieldIterator):
+                    yield item
+                else:
+                    for subitem in item:
+                        yield subitem
+
+        return LocalIterator(
+            self.base_iterator,
+            self.local_transforms + [apply_flatten],
+            name=self.name + ".flatten()")
+
+    def union(self, other: "LocalIterator[T]") -> "LocalIterator[T]":
+        """Return an iterator that is the union of this and the other.
```

Comment:
Are there ordering semantics between results from the two iterators? Or are results surfaced as they finish? Should add to comment.